src/main/java/frc/robot/command/AutoAlignCommand.java

```diff
@@ -0,0 +1,217 @@
+ The command ends when the final position is reached.
+ */
+public class AutoAlignCommand extends Command {
+  private final DriveSubsystem drive;
+  private final PositioningSubsystem positioning;
+  private final ControllerInput controller;
+
+  // Selected auto-align mode from the driver's button press
+  private AutoAlignMode selectedMode;
+
+  // Trajectories for the two phases
+  private Trajectory absoluteTrajectory;
+  private Trajectory relativeTrajectory;
+
+  // A sequential command that will follow both trajectories
+  private Command trajectoryCommand;
+
+  // Example master node positions (12 master nodes for 6 reef sides, 2 per side).
+  // These should normally be read from an external configuration (e.g. Elastic Dashboard)
+  // Here we provide dummy sample values (in meters) in field coordinates.
+  private final Pose2d[] masterNodes = new Pose2d[] {
+      new Pose2d(2.0, 7.0, Constants.ZERO_ROTATION),
+      new Pose2d(4.0, 7.0, Constants.ZERO_ROTATION),
+      new Pose2d(6.0, 7.0, Constants.ZERO_ROTATION),
+      new Pose2d(2.0, 5.0, Constants.ZERO_ROTATION),
+      new Pose2d(4.0, 5.0, Constants.ZERO_ROTATION),
+      new Pose2d(6.0, 5.0, Constants.ZERO_ROTATION),
+      new Pose2d(2.0, 3.0, Constants.ZERO_ROTATION),
+      new Pose2d(4.0, 3.0, Constants.ZERO_ROTATION),
+      new Pose2d(6.0, 3.0, Constants.ZERO_ROTATION),
+      new Pose2d(2.0, 1.0, Constants.ZERO_ROTATION),
+      new Pose2d(4.0, 1.0, Constants.ZERO_ROTATION),
+      new Pose2d(6.0, 1.0, Constants.ZERO_ROTATION)
+  };
+
+  // Mapping from auto-align mode to a relative offset for the final sub node.
+  // These offsets are translational adjustments (in meters) relative to the master node.
+  private final Map<AutoAlignMode, Translation2d> subNodeOffsets = new HashMap<>() {{
+      put(AutoAlignMode.CORAL_LEFT, new Translation2d(-0.3, 0.0));
+      put(AutoAlignMode.CORAL_RIGHT, new Translation2d(0.3, 0.0));
+      put(AutoAlignMode.ALGAE_MIDDLE, new Translation2d(0.0, 0.3));
+  }};
+
+  // Trajectory configuration taken from settings.json values (adjust as necessary)
+  private final TrajectoryConfig trajConfig = new TrajectoryConfig(3.0, 3.0)
+          .setKinematics(drive.kinematics);
+
+  public AutoAlignCommand(DriveSubsystem drive, PositioningSubsystem positioning, ControllerInput controller) {
+    this.drive = drive;
+    this.positioning = positioning;
+    this.controller = controller;
+    addRequirements(drive);
+  }
+
+  @Override
+  public void initialize() {
+    // Determine which auto-align button is pressed; assume only one is held.
+    if (controller.getAutoAlignCoralLeft()) {
+      selectedMode = AutoAlignMode.CORAL_LEFT;
+    } else if (controller.getAutoAlignCoralRight()) {
+      selectedMode = AutoAlignMode.CORAL_RIGHT;
+    } else if (controller.getAutoAlignAlgaeMiddle()) {
+      selectedMode = AutoAlignMode.ALGAE_MIDDLE;
+    } else {
+      // No button is held; cancel command.
+      cancel();
```

Comment:
Eww, don't call cancel().

If your command needs to finish, have it return `true` from `isFinished()`.

Reply:
Actually, you should probably set up this command so that the trigger cancels it when the button is released. Refer to [Trigger#whileTrue](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/wpilibj2/command/button/Trigger.html#whileTrue(edu.wpi.first.wpilibj2.command.Command))

Reply:
In fact, you probably don't even need a reference to the user input at all in this method.
Make it so that this command's job is to align to a node for as long as it is running.
Use an enum to determine which sub-node (CORAL_LEFT, CORAL_RIGHT, or ALGAE_MIDDLE) to align to at construction time.
Configure a trigger for each button that will cancel the command when released.